api/METRICS.md

```diff
@@ -86,21 +86,186 @@ group_max_from_summary(summary)
+
+  * `DemographicParity()`: for all _a_, <br>
+    |_selection_rate_(_a_) - _selection_rate_(\*)| &le; &epsilon;.
+
+  * `DemographicParity(ratio=`_r_`)`: for all _a_, <br>
+    _r_ &sdot; _selection_rate_(_a_) - _selection_rate_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _selection_rate_(\*) - _selection_rate_(_a_) &le; &epsilon;.
+
+  * `TruePositiveRateDifference()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;.
+
+  * `TruePositiveRateDifference(ratio=`_r_`)`: analogous
+
+  * `EqualizedOdds()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;, <br>
+    |_false_positive_rate_(_a_) - _false_positive_rate_(\*)| &le; &epsilon;.
+
+  * `EqualizedOdds(ratio=`_r_`)`: analogous
+
+  * `ErrorRateRatio()`: for all _a_, <br>
+    |_error_rate_(_a_) - _error_rate_(\*)| &le; &epsilon;.
+
+  * `ErrorRateDifference(ratio=`_r_`)`: analogous
+
+  * all of the above constraints are descendants of `ConditionalSelectionRate`
+
+  * the `objective` for all of the above constraints is `ErrorRate()`
+
+* `GridSearch(estimator, constraints)` considers constraints represented by the provided `Moment` object. The behavior of the `GridSearch` algorithm does not depend on the value of the right-hand side of constraints, so it is not provided to the constructor. In addition to the `Moment` objects above, `GridSearch` also supports the following:
+
+  * `GroupLossMoment(<loss>)`: for all _a_, <br>
+    _loss_(_a_) &le; &zeta;
+
+    * where `<loss>` is a _loss evaluation_ object; it supports an interface that takes `y_true` and `y_pred` as the input and returns the vector of losses evaluated on individual examples
+
+    * the `objective` for `GroupLossMoment(<loss>)` is `AverageLossMoment(<loss>)`
+
+  * both `GroupLossMoment` and `AverageLossMoment` are descendants of `ConditionalLossMoment`
+
+**Proposal.** The proposal introduces some breaking changes:
+
+* `constraints`:
+
+  * `<Metric>Parity(difference_bound=`&epsilon;`)`: for all _a_, <br>
+    |_metric_(_a_) - _metric_(\*)| &le; &epsilon;.
+
+  * `<Metric>Parity(ratio_bound=`_r_`, ratio_bound_slack=`&epsilon;`)`: for all _a_, <br>
+    _r_ &sdot; _metric_(_a_) - _metric_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _metric_(\*) - _metric_(_a_) &le; &epsilon;.
+
+  * `DemographicParity` and `EqualizedOdds` have the same calling convention as `<Metric>Parity`
+
+  * `BoundedGroupLoss(<loss>, bound=`&zeta;`)`: for all _a_, <br>
+    _loss_(_a_) &le; &zeta;
+
+    _Alternative proposals_:
+    * `LossParity(<loss>, group_max_bound=`&zeta;`)` <br>
+    * `<Loss>Parity(group_max_bound=`&zeta;`)`
+
+
+* `objective`:
+  * `<Metric>()` (for classification moments)
+
+  * `AverageLoss(<loss>)` (for loss minimization moments) <br>
+
+    _Alternative proposals_:
+    * `MeanLoss(<loss>)`
+    * `OverallLoss(<loss>)`
+    * the object `<loss>` doubles as (1) the loss evaluator and (2) the moment implementing the objective

```

Comment:
Can we foresee any situation where clubbing them together will hurt us in the future?
Why not just `LossMoment(<loss>)`? Then it's clear that the moment takes the loss evaluator. With `OverallLoss` and the others you mention it's still not entirely clear that it's a moment (at least not from the name).

Reply:
One use case where this might bite us is if we want to add support for various kinds of sample weights, e.g.:
* balanced loss (the re-weighting happens on the level of the moment, but not on the level of the loss evaluator)
* counterfactual loss (this style of re-weighting corrects for partial observability in data; again it's only incurred on the moment level, but not on the loss level)

I would prefer calling it `SomethingLoss` to better match what we do for the classification moments--where I think we will be simply calling them `<Metric>` rather than `<Metric>Moment`. Also, I don't expect that the user would ever specify this objective explicitly (it would be automatically derived from the BGL constraint object).

Reply:
The special case you mention could still be handled fairly simply with if-else, so I'm not too worried.

Hmm, then how do we distinguish between loss evaluator and loss moment in terms of names?

Reply:
It occurred to me that we may need a special loss object if we ever want to implement demographic parity for regression--and I just heard that somebody will be working on it this summer, so we definitely need both a loss evaluator object and a loss moment object.

I don't think it's necessarily a problem that we have moments called `BoundedGroupLoss` and `MeanLoss`, and loss evaluators called `SquareLoss`, `AbsoluteLoss`, and `ZeroOneLoss`. But if you feel strongly about this, we could maybe change the evaluators to `SquareLossEvaluator` etc.?